File: from-0001-to-0100/0004-median-of-two-sorted-arrays/solution.go
func median(list []int) float64 {
	mid := len(list) / 2
	if len(list)%2 == 0 {
		return float64(list[mid]+list[mid-1]) / 2.0
	}
	return float64(list[mid])
}

func findMedianSortedArrays(nums1 []int, nums2 []int) float64 {
	answer := []int{}
	for i, j := 0, 0; i < len(nums1) || j < len(nums2); {
		if i < len(nums1) && (j == len(nums2) || nums1[i] <= nums2[j]) {
			answer = append(answer, nums1[i])
			i += 1
		} else {
			answer = append(answer, nums2[j])
			j += 1
		}

	}
	return median(answer)
}

File: from-0001-to-0100/0004-median-of-two-sorted-arrays/solution.py
class Solution:
    def findMedianSortedArrays(self, nums1: List[int], nums2: List[int]) -> float:
        total = len(nums1) + len(nums2)
        half = total // 2

        if len(nums1) > len(nums2):
            nums1, nums2 = nums2, nums1

        left, right = 0, len(nums1) - 1
        while True:
            i = (left + right) // 2
            j = half - i - 2

            aleft = nums1[i] if i >= 0 else float('-inf')
            aright = nums1[i+1] if i+1 < len(nums1) else float('inf')
            bleft = nums2[j] if j >= 0 else float('-inf')
            bright = nums2[j+1] if j+1 < len(nums2) else float('inf')

            if aleft <= bright and bleft <= aright:
                if total % 2:
                    return min(aright, bright)
                return (max(aleft, bleft) + min(aright, bright)) / 2
            elif aleft > bright:
                right = i - 1
            else:
                left = i + 1
